server/src/main/java/org/opensearch/index/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,825 @@
+    }
+
+    public int ignoreAbove() {
+        return ignoreAbove;
+    }
+
+    private static final FieldType FIELD_TYPE = new FieldType();
+    static {
+        FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        FIELD_TYPE.setTokenized(true);
+        FIELD_TYPE.setOmitNorms(true);
+        FIELD_TYPE.setStored(false);
+        FIELD_TYPE.freeze();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        String value;
+        if (context.externalValueSet()) {
+            value = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                value = nullValue;
+            } else {
+                value = parser.textOrNull();
+            }
+        }
+
+        if (value == null || value.length() > ignoreAbove) {
+            return;
+        }
+
+        NamedAnalyzer normalizer = fieldType().normalizer();
+        if (normalizer != null) {
+            value = normalizeValue(normalizer, name(), value);
+        }
+
+        // convert to utf8 only once before feeding postings/dv/stored fields
+        final BytesRef binaryValue = new BytesRef(value);
+        Tokenizer tokenizer = new WildcardFieldTokenizer();
+        tokenizer.setReader(new StringReader(value));
+        context.doc().add(new Field(fieldType().name(), tokenizer, FIELD_TYPE));
+        if (fieldType().hasDocValues()) {
+            context.doc().add(new SortedSetDocValuesField(fieldType().name(), binaryValue));
+        } else {
+            if (fieldType().hasDocValues() == false) {
+                createFieldNamesField(context);
+            }
+        }
+    }
+
+    /**
+     * Tokenizer to emit tokens to support wildcard first-phase matching.
+     * <p>
+     * Will emit all substrings of length 1,2, and 3, with 0-valued anchors for the prefix/suffix.
+     * <p>
+     * For example, given the string "lucene", output the following terms:
+     * <p>
+     * [0, 'l']
+     * [0, 'l', 'u']
+     * ['l']
+     * ['l', 'u']
+     * ['l', 'u', 'c']
+     * ['u']
+     * ['u','c']
+     * ['u','c','e']
+     * ['c']
+     * ['c', 'e']
+     * ['c', 'e', 'n']
+     * ['e']
+     * ['e', 'n']
+     * ['e', 'n', 'e']
+     * ['n']
+     * ['n', 'e']
+     * ['n', 'e', 0]
+     * ['e']
+     * ['e', 0]
+     * <p>
+     * Visible for testing.
+     */
+    static final class WildcardFieldTokenizer extends Tokenizer {
+        private final CharTermAttribute charTermAttribute = addAttribute(CharTermAttribute.class);
+        private final char[] buffer = new char[3]; // Ring buffer for up to 3 chars
+        private int offset = 0; // Position in the buffer
+        private int length = 2; // First token is anchor + first char
+
+        @Override
+        public void reset() throws IOException {
+            super.reset();
+            buffer[0] = 0;
```

Comment:
if input string contains 0s, how are we differentiating them from 0s anchored to head and tail terms?

Reply:
so if someone is using string like - `"0\u0000\u0000\u0000"`, then this might produce unwanted result. Given it will not result in scenarios like infinite loop, this behaviour should be acceptable.

Reply:
I think there's an argument that the field is intended for text, so a zero byte would be considered invalid.

Maybe we should check for zero in the bytesref and fail on indexing to minimize surprises?

Reply:
I think the current behaviour is fine given the likelihood of it happening is low and the consequence is not where its runs into some failure or its producing extra unwanted search results. It will be skipping these documents from search results as there will be less number of tokens produced while indexing/search than desired.